### PR TITLE
ddl: fix rawArgs is overwritten

### PR DESCRIPTION
**Pull request by zimulala**
We run DDL job in the case as follows:
we encounter an error before calling DecodeArgs，and we don’t cancel this DDL job, so we will use the current job(with error info) to overwrite the old one. But the current job’s Args is nil(because this case returns the error before DecodeArgs).
	The error mentioned earlier is like `TiKV server timeout[try again later]` when we get table information from TiKV.

### CODE REVIEW

**Comment by shenli** [comment]
@coocood @winkyao PTAL

File: ddl/ddl_worker.go [hunk 1]
```diff
@@ -123,10 +123,16 @@ func (d *ddl) handleUpdateJobError(t *meta.Meta, job *model.Job, err error) erro
...
 // Every time we enter another state except final state, we must call this function.
-func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64) error {
+func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64, err error) error {
 	job.LastUpdateTS = int64(updateTS)
-	err := t.UpdateDDLJob(0, job)
-	return errors.Trace(err)
+	updateRawArgs := true
+	// If there is an error when running job and the RawArgs hasn't been decoded wit DecodeArgs,
+	// so we should't replace RawArgs with the marshaling Args.
+	if err != nil && (job.RawArgs != nil && job.Args == nil) {
```
Comment (winkyao) [inline ddl/ddl_worker.go]: why not just use `job.Error != nil`?

**Comment by coocood** [comment]
@zimulala 
There is three places `job.Encode` is called.
1. `enQueueDDLJob`
2. `updateDDLJob`
3. `addHistoryDDLJob`

Currently, all the 3 places use true.
I think we can always use false in `updateDDLJob`.

**Comment by coocood** [comment]
@zimulala 
Never mind, sometimes we have to update the raw args.

**Comment by coocood** [comment]
LGTM

File: ddl/ddl_worker.go [hunk 2]
```diff
@@ -123,10 +123,16 @@ func (d *ddl) handleUpdateJobError(t *meta.Meta, job *model.Job, err error) erro
...
 // updateDDLJob updates the DDL job information.
 // Every time we enter another state except final state, we must call this function.
-func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64) error {
+func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64, err error) error {
 	job.LastUpdateTS = int64(updateTS)
-	err := t.UpdateDDLJob(0, job)
-	return errors.Trace(err)
+	updateRawArgs := true
+	// If there is an error when running job and the RawArgs hasn't been decoded wit DecodeArgs,
+	// so we should't replace RawArgs with the marshaling Args.
```
Comment (coocood) [inline ddl/ddl_worker.go]: s/should't/shouldn't

File: ddl/ddl_worker.go [hunk 3]
```diff
@@ -123,10 +123,16 @@ func (d *ddl) handleUpdateJobError(t *meta.Meta, job *model.Job, err error) erro
 
 // updateDDLJob updates the DDL job information.
 // Every time we enter another state except final state, we must call this function.
-func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64) error {
+func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64, err error) error {
 	job.LastUpdateTS = int64(updateTS)
-	err := t.UpdateDDLJob(0, job)
-	return errors.Trace(err)
+	updateRawArgs := true
+	// If there is an error when running job and the RawArgs hasn't been decoded wit DecodeArgs,
```
Comment (coocood) [inline ddl/ddl_worker.go]: s/wit/by

File: ddl/ddl_worker.go (hunk 1 above)
Comment (zimulala) [reply ddl/ddl_worker.go]: Because `job.Error` will keep the previous error.

**Comment by jackysp** [comment]
Will there be some integration tests to cover this case?

**Comment by zimulala** [comment]
@jackysp Now I added tests.

**Comment by zimulala** [comment]
PTAL @winkyao @shenli @jackysp

File: ddl/ddl_worker.go [hunk 4]
```diff
@@ -123,10 +123,16 @@ func (d *ddl) handleUpdateJobError(t *meta.Meta, job *model.Job, err error) erro
 
 // updateDDLJob updates the DDL job information.
 // Every time we enter another state except final state, we must call this function.
-func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64) error {
+func (d *ddl) updateDDLJob(t *meta.Meta, job *model.Job, updateTS uint64, err error) error {
```
Comment (shenli) [inline ddl/ddl_worker.go]: The last argument is strange.

**Comment by shenli** [comment]
LGTM

**Pull request merged**